minio/minio
# Concurrent  CSV parsing and reduce S3 select allocations

## Description

Make CSV decoding concurrent.

Decode 128KB blocks in separate goroutines.

Remove 2 of 3 allocations from CSV record reads and do cheaper stream transformation.

Benchmarks on sizes ~100K to 100MB. 6 cores:
```
BenchmarkReaderHuge/97K-12         	    2367	    503932 ns/op	 197.82 MB/s	  818116 B/op	     694 allocs/op
BenchmarkReaderHuge/194K-12        	    1562	    712566 ns/op	 279.80 MB/s	 1356900 B/op	    1337 allocs/op
BenchmarkReaderHuge/389K-12        	    1227	    897329 ns/op	 444.38 MB/s	 2435629 B/op	    2634 allocs/op
BenchmarkReaderHuge/778K-12        	     796	   1445879 ns/op	 551.58 MB/s	 4411183 B/op	    5182 allocs/op
BenchmarkReaderHuge/1557K-12       	     458	   2536879 ns/op	 628.73 MB/s	 8307241 B/op	   10226 allocs/op
BenchmarkReaderHuge/3115K-12       	     289	   4054895 ns/op	 786.72 MB/s	11528289 B/op	   16071 allocs/op
BenchmarkReaderHuge/6230K-12       	     170	   7110386 ns/op	 897.29 MB/s	15152059 B/op	   26810 allocs/op
BenchmarkReaderHuge/12461K-12      	     100	  11888191 ns/op	1073.35 MB/s	21751391 B/op	   48029 allocs/op
BenchmarkReaderHuge/24922K-12      	      62	  20445297 ns/op	1248.23 MB/s	34399005 B/op	   90288 allocs/op
BenchmarkReaderHuge/49844K-12      	      34	  38455932 ns/op	1327.25 MB/s	59573994 B/op	  174408 allocs/op
BenchmarkReaderHuge/99689K-12      	      16	  71184500 ns/op	1434.04 MB/s	109693871 B/op	  343077 allocs/op
PASS
```

At files at 200KB we have the same speed as current implementation.

If we want more speed, we could at one point rewrite the CSV parser to operate purely on byte slices and make it practically allocation-less.

## Motivation and Context

Faster S3 select throughput.

## How to test this PR?

We probably should add more diverse tests to the parsing, however this doesn't change much in terms 

## Types of changes

## Checklist:
